snapshot-labs/sx-monorepo
# feat: enable voting from Safe

### Summary
Closes #263 

- Updated manifest.json to include icon and imported manifest in index.html (used for safe app)
- New composable `useSafeWallet.ts` to check if wallet connected is Safe wallet or not, Also checks if safe wallet is using same network as space
- Show error on Proposal page if Safe wallet's network is different from safe network 
- Should be able to vote off-chain with Safe wallet (both from safe app and wallet connect)
- Should be able to vote on-chain with Safe wallet (both from safe app and wallet connect)
- If signature is `0x` UI will  show a notification that vote is pending and waiting for other signers


### How to test
- #### Add Snapshot as a custom Safe App (You can also connect with safe's wallet-connect)
  - Go to your Safe -> Apps -> My Custom Apps -> Click on "Add Custom App" button
  - Paste the snapshot UI url, for example https://bafybeihf65ituuynsylpwh7s4nj7txfhy27d7nlbi2jhg3wigze6rr2i2y.on.fleek.co/#/
  - You will see following App card if everything is correct
  - ![image](https://github.com/snapshot-labs/sx-monorepo/assets/15967809/f0d611fd-7a1c-4ea9-a7f9-08d98b7d02ed) 
  - Once you click on "Add" you will see that app is added to the list 
  - ![image](https://github.com/snapshot-labs/sx-monorepo/assets/15967809/fc0fa65e-f53a-4591-99fe-91201bb9cff5)
  - Navigate to the proposal you want to vote
 
- #### Choosing between on-chain and off-chain with Safe wallet:
  - By default safe wallet will use off-chain, to enable on-chain, refer to: https://docs.snapshot.org/user-guides/using-safe-multi-sig#asynchronous-signing

 
- #### Try to vote
- If you are connected to a safe that is on different network than space's network, then you will see following error
- ![image](https://github.com/snapshot-labs/sx-monorepo/assets/15967809/8158da2e-8c9d-4670-903d-c084e41eabe6)
- Try voting off-chain and on-chain
- Try voting with a Safe that require one confirmation
- Try voting with a Safe that require multiple confirmations


### To-Do
- [ ] Before merging remove this https://github.com/snapshot-labs/sx-monorepo/blob/enable-voting-from-safe/apps/ui/src/views/Home.vue#L42-L46
